storage/isolation_manager.rs

```diff
@@ -40,130 +43,256 @@ pub(crate) struct IsolationManager {
 
 impl fmt::Display for IsolationManager {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(
-            f,
-            "Timeline[windows={}, next_window_sequence_number={}, watermark={}]",
-            self.timeline.window_count(),
-            self.timeline.next_window_sequence_number(),
-            self.watermark()
-        )
+        write!(f, "Timeline[windows={}, watermark={}]", self.timeline.window_count(), self.watermark())
     }
 }
 
 impl IsolationManager {
-    pub(crate) fn new(next_sequence_number: SequenceNumber) -> IsolationManager {
-        IsolationManager { timeline: Timeline::new(next_sequence_number) }
+    pub(crate) fn new(next_relative_index: i64) -> IsolationManager {
+        IsolationManager { timeline: Timeline::new(next_relative_index) }
     }
 
     pub(crate) fn opened(&self, open_sequence_number: SequenceNumber) {
+        debug_assert!(open_sequence_number <= self.watermark());
         self.timeline.record_reader(open_sequence_number);
     }
 
     pub(crate) fn closed(&self, open_sequence_number: SequenceNumber) {
         self.timeline.remove_reader(open_sequence_number);
     }
 
-    pub(crate) fn try_commit(
+    pub(crate) fn load_applied(
         &self,
-        commit_sequence_number: SequenceNumber,
+        relative_index: i64,
+        sequence_number: SequenceNumber,
         commit_record: CommitRecord,
-    ) -> Result<(), IsolationError> {
-        let open_sequence_number = commit_record.open_sequence_number;
-        let commit_window = self.pending(commit_sequence_number, commit_record);
-        let validation_result =
-            self.validate_all_concurrent(commit_sequence_number, commit_window.get_record(commit_sequence_number));
-
-        match &validation_result {
-            Ok(()) => self.committed(commit_sequence_number),
-            Err(_) => self.commit_failed(commit_sequence_number, open_sequence_number),
-        }
+    ) {
+        let (window, slot_index) = self.timeline.get_or_create_window(relative_index);
+        window.insert_pending(slot_index, sequence_number, commit_record);
+        window.set_applied(slot_index);
+        self.timeline.may_increment_watermark(relative_index);
+    }
 
-        validation_result
+    pub(crate) fn load_aborted(&self, relative_index: i64, sequence_number: SequenceNumber) {
+        let (window, slot_index) = self.timeline.get_or_create_window(relative_index);
+        window.insert_pending(slot_index, sequence_number, CommitRecord::new(KeyspaceBuffers::new(), sequence_number));
+        window.set_closed(slot_index);
```

Comment:
If we call this method `set_aborted` (see comments in `storage.rs`) we shouldn't need to insert pending and then set closed immediately right?
Also, the terminology sohuld be consistent - `closed` should become `aborted` here too right?

Reply:
We need to insert pending to initialise the OnceLock and write the sequence number.